File: desktop/BcmDiagnosticTool/Protocol/FrameCodec.cs
namespace BcmDiagnosticTool.Protocol;

/// <summary>
/// Encodes frames onto the wire, per BCM-ICD-001 section 3.
/// </summary>
public static class FrameCodec
{
    /// <summary>
    /// Build the on-the-wire bytes for a request.
    /// </summary>
    /// <exception cref="ArgumentException">payload longer than 32 bytes.</exception>
    public static byte[] Encode(byte cmd, ReadOnlySpan<byte> payload)
    {
        if (payload.Length > Icd.MaxPayload)
        {
            throw new ArgumentException(
                $"payload is {payload.Length} bytes, maximum is {Icd.MaxPayload}",
                nameof(payload));
        }

        var frame = new byte[payload.Length + Icd.FrameOverhead];
        int i = 0;

        frame[i++] = Icd.Header;
        frame[i++] = cmd;
        frame[i++] = (byte)payload.Length;
        foreach (byte b in payload) { frame[i++] = b; }

        // CRC covers CMD, LEN and PAYLOAD - not the header or the footer.
        frame[i++] = Crc8.Compute(frame.AsSpan(1, payload.Length + 2));
        frame[i] = Icd.Footer;

        return frame;
    }

    public static byte[] Encode(Cmd cmd, params byte[] payload) =>
        Encode((byte)cmd, payload);
}

/// <summary>
/// Incremental byte-at-a-time parser, mirroring the firmware's state machine.
///
/// On a bad checksum or a missing footer it reports the error and resumes
/// hunting for the next header, so a corrupted frame costs one frame rather
/// than desynchronising the link.
/// </summary>
public sealed class FrameParser
{
    public enum Result
    {
        NeedMore,
        Complete,
        ChecksumError,
        FramingError,
        Overflow,
    }

    private enum State { Header, Command, Length, Payload, Checksum, Footer }

    private State _state = State.Header;
    private byte _cmd;
    private byte _len;
    private byte[] _payload = Array.Empty<byte>();
    private int _index;
    private readonly List<byte> _crcInput = new();

    public Frame? Frame { get; private set; }
    public uint FramesOk { get; private set; }
    public uint FramesBad { get; private set; }

    public void Reset()
    {
        _state = State.Header;
        _index = 0;
        _crcInput.Clear();
    }

    private Result Fail(Result why)
    {
        FramesBad++;
        Reset();
        return why;
    }

    public Result Feed(byte b)
    {
        switch (_state)
        {
            case State.Header:
                // Anything that is not a header is discarded. This is how the
                // parser recovers from corruption or a mid-stream connect.
                if (b == Icd.Header)
                {
                    _crcInput.Clear();
                    _index = 0;
                    _state = State.Command;
                }
                return Result.NeedMore;

            case State.Command:
                _cmd = b;
                _crcInput.Add(b);
                _state = State.Length;
                return Result.NeedMore;

            case State.Length:
                if (b > Icd.MaxPayload)
                {
                    // A bogus length would make us swallow arbitrary bytes.
                    return Fail(Result.Overflow);
                }
                _len = b;
                _crcInput.Add(b);
                _payload = new byte[_len];
                _index = 0;
                _state = _len == 0 ? State.Checksum : State.Payload;
                return Result.NeedMore;

            case State.Payload:
                _payload[_index++] = b;
                _crcInput.Add(b);
                if (_index >= _len) { _state = State.Checksum; }
                return Result.NeedMore;

            case State.Checksum:
                if (b != Crc8.Compute(System.Runtime.InteropServices
                                            .CollectionsMarshal.AsSpan(_crcInput)))
                {
                    return Fail(Result.ChecksumError);
                }
                _state = State.Footer;
                return Result.NeedMore;

            case State.Footer:
                if (b != Icd.Footer) { return Fail(Result.FramingError); }
                Frame = new Frame { Cmd = _cmd, Payload = _payload };
                FramesOk++;
                Reset();
                return Result.Complete;

            default:
                return Fail(Result.FramingError);
        }
    }
}

File: firmware/services/frame_codec.cpp
/**
 * @file    frame_codec.cpp
 * @brief   Frame encoding and byte-stream parsing.
 */
#include "frame_codec.h"

namespace bcm {
namespace services {

bool FrameCodec::encode(const Frame& frame,
                        uint8_t*     out,
                        uint16_t     out_size,
                        uint16_t&    written)
{
    written = 0U;

    if (out == 0)                     { return false; }
    if (frame.len > kMaxPayload)      { return false; }

    const uint16_t need = encoded_size(frame);
    if (out_size < need)              { return false; }

    Crc8     crc;
    uint16_t i = 0U;

    out[i++] = kFrameHeader;

    out[i++] = frame.cmd;
    crc.update(frame.cmd);

    out[i++] = frame.len;
    crc.update(frame.len);

    for (uint8_t p = 0U; p < frame.len; ++p) {
        out[i++] = frame.payload[p];
        crc.update(frame.payload[p]);
    }

    out[i++] = crc.value();
    out[i++] = kFrameFooter;

    written = i;
    return true;
}

void FrameParser::reset()
{
    state_ = State::Header;
    index_ = 0U;
    crc_.reset();
}

FrameParser::Result FrameParser::fail(Result why)
{
    ++frames_bad_;
    reset();
    return why;
}

FrameParser::Result FrameParser::feed(uint8_t byte)
{
    switch (state_) {

    case State::Header:
        /* Silently discard anything that is not a header - this is how the
         * parser recovers from a corrupted frame or a mid-stream connect. */
        if (byte == kFrameHeader) {
            crc_.reset();
            frame_.cmd = 0U;
            frame_.len = 0U;
            index_     = 0U;
            state_     = State::Command;
        }
        return Result::NeedMore;

    case State::Command:
        frame_.cmd = byte;
        crc_.update(byte);
        state_ = State::Length;
        return Result::NeedMore;

    case State::Length:
        if (byte > kMaxPayload) {
            /* A bogus length would otherwise make us swallow arbitrary
             * bytes, so reject immediately and resynchronise. */
            return fail(Result::Overflow);
        }
        frame_.len = byte;
        crc_.update(byte);
        index_ = 0U;
        state_ = (byte == 0U) ? State::Checksum : State::Payload;
        return Result::NeedMore;

    case State::Payload:
        frame_.payload[index_] = byte;
        crc_.update(byte);
        ++index_;
        if (index_ >= frame_.len) { state_ = State::Checksum; }
        return Result::NeedMore;

    case State::Checksum:
        if (byte != crc_.value()) { return fail(Result::ChecksumError); }
        state_ = State::Footer;
        return Result::NeedMore;

    case State::Footer:
        if (byte != kFrameFooter) { return fail(Result::FramingError); }
        ++frames_ok_;
        state_ = State::Header;
        index_ = 0U;
        return Result::Complete;

    default:
        return fail(Result::FramingError);
    }
}

}  // namespace services
}  // namespace bcm
